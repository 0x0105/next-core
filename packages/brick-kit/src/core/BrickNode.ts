import {
  PluginRuntimeContext,
  BrickLifeCycle,
  RefForProxy,
  CustomTemplateProxy,
} from "@easyops/brick-types";
import { bindListeners } from "../bindListeners";
import { setRealProperties } from "../setProperties";
import { handleProxyOfCustomTemplate } from "./exports";

export interface RuntimeBrick {
  type?: string;
  properties?: Record<string, any>;
  template?: string;
  params?: Record<string, any>;
  events?: Record<string, any>;
  children?: RuntimeBrick[];
  slotId?: string;
  context?: PluginRuntimeContext;
  lifeCycle?: BrickLifeCycle;
  element?: HTMLElement;
  bg?: boolean;
  proxy?: CustomTemplateProxy;
  proxyRefs?: Map<string, RefForProxy>;
  refForProxy?: {
    brick?: RuntimeBrick;
  };
}

export class BrickNode {
  private currentElement: RuntimeBrick;
  private children: BrickNode[];

  constructor(brick: RuntimeBrick) {
    this.currentElement = brick;
  }

  mount(): HTMLElement {
    const brick = this.currentElement;
    const tagName = brick.type;

    if (tagName.includes("-") && !customElements.get(tagName)) {
      // eslint-disable-next-line no-console
      console.error(`Undefined custom element: ${tagName}`);
    }

    if (tagName === "basic-bricks.script-brick") {
      // eslint-disable-next-line no-console
      console.warn(
        "`basic-bricks.script-brick` is deprecated, please take caution when using it"
      );
    }

    const node = document.createElement(tagName);
    brick.element = node;

    if (brick.slotId) {
      node.setAttribute("slot", brick.slotId);
    }
    setRealProperties(node, brick.properties);
    // Todo(steve): refine
    bindListeners(node, brick.events, brick.context);

    if (Array.isArray(brick.children)) {
      this.children = brick.children.map((slot) => new BrickNode(slot));
      const childNodes = this.children.map((slot) => slot.mount());
      childNodes.forEach((slot) => node.appendChild(slot));
    } else {
      this.children = [];
    }

    handleProxyOfCustomTemplate(brick);

    return node;
  }

  unmount(): void {
    this.children.forEach((slot) => {
      slot.unmount();
    });
  }
}
