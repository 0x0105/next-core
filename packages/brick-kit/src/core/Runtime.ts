import {
  MountPoints,
  MicroApp,
  InterceptorParams,
  FeatureFlags,
  DesktopData,
  UserInfo,
  BrickPackage,
  Storyboard,
  MagicBrickConfig,
  PluginRuntimeContext,
  BrickConf,
  TemplatePackage,
} from "@easyops/brick-types";
import {
  Kernel,
  MenuBar,
  AppBar,
  Resolver,
  registerCustomTemplate,
  registerCustomProcessor,
} from "./exports";
import { registerBrickTemplate } from "./TemplateRegistries";
import {
  RelatedApp,
  RouterState,
  RecentApps,
  CustomApiOrchestration,
  AbstractRuntime,
} from "./interfaces";

let kernel: Kernel;

/* istanbul ignore next */
export function _dev_only_getBrickPackages(): BrickPackage[] {
  return kernel.bootstrapData.brickPackages;
}

/* istanbul ignore next */
export function _dev_only_getTemplatePackages(): TemplatePackage[] {
  return kernel.bootstrapData.templatePackages;
}

/* istanbul ignore next */
export function _dev_only_getStoryboards(): Storyboard[] {
  // Deprecated since the storyboards are probably not fulfilled.
  return [];
}

/* istanbul ignore next */
export function _dev_only_loadDynamicBricksInBrickConf(
  brickConf: BrickConf
): Promise<void> {
  return kernel.loadDynamicBricksInBrickConf(brickConf);
}

export class Runtime implements AbstractRuntime {
  async bootstrap(mountPoints: MountPoints): Promise<void> {
    if (kernel !== undefined) {
      throw new Error("Cannot bootstrap more than once.");
    }
    kernel = new Kernel();
    await kernel.bootstrap(mountPoints);
  }

  get menuBar(): MenuBar {
    return kernel.menuBar;
  }

  get appBar(): AppBar {
    return kernel.appBar;
  }

  /* istanbul ignore next */
  getCurrentApp(): MicroApp {
    return kernel.currentApp;
  }

  /* istanbul ignore next */
  getRecentApps(): RecentApps {
    return kernel.getRecentApps();
  }

  getMicroApps({
    excludeInstalling = false,
    includeInternal = false,
  } = {}): MicroApp[] {
    let apps = kernel.bootstrapData.microApps;
    if (excludeInstalling) {
      apps = apps.filter(
        (app) => !(app.installStatus && app.installStatus === "running")
      );
    }
    if (!includeInternal) {
      apps = apps.filter((app) => !app.internal);
    }
    return apps;
  }

  reloadMicroApps(interceptorParams?: InterceptorParams): Promise<void> {
    return kernel.loadMicroApps(
      {
        check_login: true,
      },
      interceptorParams
    );
  }

  /* istanbul ignore next */
  reloadSharedData(): void {
    return kernel.loadSharedData();
  }

  /* istanbul ignore next */
  getDesktops(): DesktopData[] {
    return kernel.bootstrapData.desktops || [];
  }

  /* istanbul ignore next */
  getAllUserInfo(): UserInfo[] {
    // eslint-disable-next-line no-console
    console.warn(
      "`getRuntime().getAllUserInfo()` is deprecated and will always return an empty array, please use `await getRuntime().getAllUserMapAsync()` instead"
    );
    return [];
  }

  /* istanbul ignore next */
  getAllUserMap(): Map<string, UserInfo> {
    // eslint-disable-next-line no-console
    console.warn(
      "`getRuntime().getAllUserMap()` is deprecated and will always return an empty Map, please use `await getRuntime().getAllUserMapAsync()` instead"
    );
    return new Map();
  }

  /* istanbul ignore next */
  getAllUserMapAsync(): Promise<Map<string, UserInfo>> {
    return kernel.allUserMapPromise;
  }

  /* istanbul ignore next */
  getMagicBrickConfigMapAsync(): Promise<Map<string, MagicBrickConfig>> {
    return kernel.allMagicBrickConfigMapPromise;
  }

  /**
   * 切换主体内容 `filter: blur(...)`;
   * @deprecated
   * @param blur
   */
  toggleFilterOfBlur(blur: boolean): void {
    document.body.classList.toggle("filter-of-blur", blur);
  }

  toggleLaunchpadEffect(open: boolean): void {
    document.body.classList.toggle("launchpad-open", open);
  }

  /* istanbul ignore next */
  getFeatureFlags(): FeatureFlags {
    return kernel.getFeatureFlags();
  }

  getHomepage(): string {
    return kernel.bootstrapData.settings?.homepage ?? "/";
  }

  getBrandSettings(): Record<string, string> {
    return Object.assign(
      { base_title: "DevOps 管理专家" },
      kernel.bootstrapData.settings?.brand
    );
  }

  getLaunchpadSettings(): { columns: number; rows: number } {
    return Object.assign(
      {
        columns: 7,
        rows: 4,
      },
      kernel.bootstrapData.settings?.launchpad
    );
  }

  getMiscSettings(): Record<string, unknown> {
    return Object.assign({}, kernel.bootstrapData.settings?.misc);
  }

  registerBrickTemplate = registerBrickTemplate;
  registerCustomTemplate = registerCustomTemplate;
  registerCustomProcessor = registerCustomProcessor;

  /* istanbul ignore next */
  getRelatedApps(appId: string): RelatedApp[] {
    // eslint-disable-next-line no-console
    console.warn(
      "`getRuntime().getRelatedApps()` is deprecated and will always return an empty array, please use `await getRuntime().getRelatedAppsAsync()` instead"
    );
    return [];
  }

  /* istanbul ignore next */
  getRelatedAppsAsync(appId: string): Promise<RelatedApp[]> {
    return kernel.getRelatedAppsAsync(appId);
  }

  /* istanbul ignore next */
  popWorkspaceStack(): void {
    return kernel.popWorkspaceStack();
  }

  /* istanbul ignore next */
  resetWorkspaceStack(): void {
    kernel.workspaceStack = [];
  }
}

/* istanbul ignore next */
export function _internalApiGetResolver(): Resolver {
  return kernel.router.getResolver();
}

/* istanbul ignore next */
export function _internalApiGetRouterState(): RouterState {
  return kernel.router.getState();
}

/* istanbul ignore next */
export function _internalApiMessageCloseHandler(event: CloseEvent): void {
  return kernel.router.handleMessageClose(event);
}

/* istanbul ignore next */
export function _internalApiGetCurrentContext(): PluginRuntimeContext {
  if (process.env.NODE_ENV === "test") {
    return {} as any;
  }
  return kernel.router.getCurrentContext();
}

/* istanbul ignore next */
export function _internalApiGetProviderBrick(
  provider: string
): Promise<HTMLElement> {
  return kernel.getProviderBrick(provider);
}

/* istanbul ignore next */
export function _internalApiGetMicroAppApiOrchestrationMap(): Promise<
  Map<string, CustomApiOrchestration>
> {
  return kernel.allMicroAppApiOrchestrationPromise;
}
