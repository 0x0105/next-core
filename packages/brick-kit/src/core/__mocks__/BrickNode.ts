import { RuntimeBrick } from "../BrickNode";

export class BrickNode {
  private children;
  constructor(private brick: RuntimeBrick) {}

  mount(): HTMLElement {
    return document.createElement(this.brick.type);
  }

  unmount(): void {}
}
