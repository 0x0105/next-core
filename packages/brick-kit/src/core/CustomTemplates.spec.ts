import {
  expandCustomTemplate,
  registerCustomTemplate,
  getTagNameOfCustomTemplate,
  handleProxyOfCustomTemplate,
} from "./CustomTemplates";
import { RuntimeBrick } from "./BrickNode";
import * as runtime from "./Runtime";

jest.spyOn(runtime, "_internalApiGetCurrentContext").mockReturnValue({} as any);

describe("expandCustomTemplate", () => {
  beforeAll(() => {
    registerCustomTemplate("steve-test.custom-template", {
      proxy: {
        properties: {
          button: {
            ref: "button",
            refProperty: "buttonName",
          },
          noGap: {
            ref: "micro-view",
            refProperty: "noGap",
          },
          isDanger: {
            ref: "button",
            refTransform: {
              buttonType: "<% DATA.isDanger ? 'danger' : 'default' %>",
              style: {
                display: "<% DATA.isDanger ? 'inline' : 'block' %>",
              },
            },
          },
        },
        events: {
          "button.click": {
            ref: "button",
            refEvent: "general.button.click",
          },
        },
        slots: {
          tools: {
            ref: "micro-view",
            refSlot: "toolbar",
            refPosition: 0,
          },
          extraContent: {
            ref: "micro-view",
            refSlot: "content",
          },
        },
        methods: {
          tell: {
            ref: "button",
            refMethod: "tellStory",
          },
        },
      },
      bricks: [
        {
          brick: "basic-bricks.micro-view",
          ref: "micro-view",
          properties: {
            pageTitle: "Testing Template",
          },
          slots: {
            content: {
              type: "bricks",
              bricks: [
                {
                  brick: "basic-bricks.general-button",
                  ref: "button",
                  properties: {
                    buttonType: "dashed",
                  },
                  events: {
                    "general.button.click": {
                      action: "console.log",
                      args: ["source", "${EVENT}"],
                    },
                  },
                },
                {
                  brick: "basic-bricks.brick-in-portal",
                  portal: true,
                },
              ],
            },
          },
        },
      ],
    });
  });

  it("should define a custom element", () => {
    const tpl = customElements.get("steve-test.custom-template");
    expect((tpl as any)._dev_only_definedProperties).toEqual([
      "button",
      "noGap",
      "isDanger",
    ]);
    expect(tpl.prototype.$$typeof).toBe("custom-template");
  });

  it("should work for getTagNameOfCustomTemplate", () => {
    expect(getTagNameOfCustomTemplate("steve-test.custom-template")).toBe(
      "steve-test.custom-template"
    );
    expect(getTagNameOfCustomTemplate("custom-template", "steve-test")).toBe(
      "steve-test.custom-template"
    );
    expect(getTagNameOfCustomTemplate("steve-test.another-template")).toBe(
      false
    );
  });

  it("should expandCustomTemplate", () => {
    const proxyBrick: RuntimeBrick = {};
    const expanded = expandCustomTemplate(
      {
        brick: "steve-test.custom-template",
        properties: {
          noGap: "${QUERY.gap|bool|not}",
        },
        lifeCycle: {
          useResolves: [
            {
              provider: "providers-of-cmdb\\.cmdb-object-api-get-detail",
              args: ["HOST"],
              transform: {
                button: "@{name}",
              },
            },
          ],
        },
        events: {
          "button.click": {
            action: "console.log",
            args: ["proxied", "${EVENT}"],
          },
        },
        slots: {
          tools: {
            type: "bricks",
            bricks: [
              {
                brick: "div",
                properties: {
                  textContent: "slotted tools!",
                },
              },
            ],
          },
          extraContent: {
            type: "bricks",
            bricks: [
              {
                brick: "div",
                properties: {
                  textContent: "slotted extra content!",
                },
              },
            ],
          },
        },
      },
      proxyBrick
    );

    expect(expanded).toMatchSnapshot();
    expect(proxyBrick).toMatchSnapshot();
  });
});

describe("handleProxyOfCustomTemplate", () => {
  it("should handleProxyOfCustomTemplate for no proxy", () => {
    const getElement = (): HTMLElement => document.createElement("div");
    expect(() => {
      handleProxyOfCustomTemplate({
        element: getElement(),
      });
      handleProxyOfCustomTemplate({
        element: getElement(),
        proxy: {},
      });
      handleProxyOfCustomTemplate({
        element: getElement(),
        proxyRefs: new Map(),
      });
      handleProxyOfCustomTemplate({
        element: getElement(),
        proxy: {},
        proxyRefs: new Map(),
      });
      handleProxyOfCustomTemplate({
        element: getElement(),
        proxy: {
          properties: {
            button: {
              ref: "button",
              refProperty: "buttonName",
            },
            noGap: {
              ref: "micro-view",
              refProperty: "noGap",
            },
          },
          events: {
            "button.click": {
              ref: "button",
              refEvent: "general.button.click",
            },
          },
          methods: {
            tell: {
              ref: "button",
              refMethod: "tellStory",
            },
          },
        },
        proxyRefs: new Map(),
      });
    }).not.toThrow();
  });

  it("should handleProxyOfCustomTemplate", async () => {
    const tplElement = document.createElement("div") as any;
    const button = document.createElement("div") as any;
    const microView = document.createElement("div") as any;
    tplElement.appendChild(button);
    tplElement.appendChild(microView);

    button.buttonName = "original button name";
    button.buttonType = "default";
    button.style.display = "block";
    button.tellStory = jest.fn();
    microView.noGap = false;
    const nonBubbleEvents: CustomEvent[] = [];
    tplElement.addEventListener("button.click", (event: CustomEvent) => {
      nonBubbleEvents.push(event);
    });
    const bubbleEvents: CustomEvent[] = [];
    tplElement.addEventListener("bubbles.happen", (event: CustomEvent) => {
      bubbleEvents.push(event);
    });

    const brick: RuntimeBrick = {
      element: tplElement,
      proxy: {
        properties: {
          button: {
            ref: "button",
            refProperty: "buttonName",
          },
          noGap: {
            ref: "micro-view",
            refProperty: "noGap",
          },
          isDanger: {
            ref: "button",
            refTransform: {
              buttonType: "<% DATA.isDanger ? 'danger' : 'default' %>",
              style: {
                display: "<% DATA.isDanger ? 'inline' : 'block' %>",
              },
            },
          },
        },
        events: {
          "button.click": {
            ref: "button",
            refEvent: "general.button.click",
          },
          "bubbles.happen": {
            ref: "button",
            refEvent: "bubbles.happen",
          },
        },
        methods: {
          tell: {
            ref: "button",
            refMethod: "tellStory",
          },
        },
      },
      proxyRefs: new Map([
        [
          "button",
          {
            brick: {
              element: button,
            },
          },
        ],
        [
          "micro-view",
          {
            brick: {
              element: microView,
            },
          },
        ],
      ]),
    };
    handleProxyOfCustomTemplate(brick);

    // Changed prop in proxy.
    expect(tplElement.button).toEqual("original button name");
    tplElement.button = "new button name";
    expect(tplElement.button).toEqual("new button name");
    expect(button.buttonName).toEqual("new button name");

    // Same prop name as proxy.
    expect(tplElement.noGap).toEqual(false);
    tplElement.noGap = true;
    expect(tplElement.noGap).toEqual(true);
    expect(microView.noGap).toEqual(true);

    // Changed transformable prop in proxy.
    expect(tplElement.isDanger).toEqual(undefined);
    tplElement.isDanger = true;
    expect(tplElement.isDanger).toEqual(true);
    expect(button.buttonType).toEqual("danger");
    expect(button.style.display).toEqual("inline");

    // Invoke a method.
    tplElement.tell("good", "story");
    expect(button.tellStory).toBeCalledWith("good", "story");

    // Dispatch an event.
    button.dispatchEvent(
      new CustomEvent("general.button.click", {
        detail: "oops!",
        cancelable: true,
      })
    );
    expect(nonBubbleEvents.length).toBe(1);
    expect(nonBubbleEvents[0].type).toBe("button.click");
    expect(nonBubbleEvents[0].detail).toBe("oops!");
    expect(nonBubbleEvents[0].cancelable).toBe(true);
    expect(nonBubbleEvents[0].bubbles).toBe(false);
    expect(nonBubbleEvents[0].composed).toBe(false);

    // Dispatch an bubble event.
    button.dispatchEvent(
      new CustomEvent("bubbles.happen", {
        detail: "bubbles!",
        bubbles: true,
      })
    );
    expect(bubbleEvents.length).toBe(1);
    expect(bubbleEvents[0].type).toBe("bubbles.happen");
    expect(bubbleEvents[0].detail).toBe("bubbles!");
    expect(bubbleEvents[0].cancelable).toBe(false);
    expect(bubbleEvents[0].bubbles).toBe(true);
    expect(bubbleEvents[0].composed).toBe(false);

    expect(tplElement.$$getElementByRef("button")).toBe(button);
    expect(tplElement.$$getElementByRef("not-existed")).toBe(undefined);
  });
});
