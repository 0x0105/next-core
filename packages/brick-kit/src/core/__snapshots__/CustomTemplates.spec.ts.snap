// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expandCustomTemplate should expandCustomTemplate 1`] = `
Object {
  "brick": "steve-test.custom-template",
  "events": Object {
    "button.click": Object {
      "action": "console.log",
      "args": Array [
        "proxied",
        "\${EVENT}",
      ],
    },
  },
  "lifeCycle": Object {
    "useResolves": Array [
      Object {
        "args": Array [
          "HOST",
        ],
        "provider": "providers-of-cmdb\\\\.cmdb-object-api-get-detail",
        "transform": Object {
          "button": "@{name}",
        },
      },
    ],
  },
  "slots": Object {
    "": Object {
      "bricks": Array [
        Object {
          "$$computedPropsFromProxy": Object {
            "noGap": "\${QUERY.gap|bool|not}",
          },
          "$$parentTemplate": undefined,
          "$$refForProxy": Object {},
          "brick": "basic-bricks.micro-view",
          "properties": Object {
            "pageTitle": "Testing Template",
          },
          "slots": Object {
            "content": Object {
              "bricks": Array [
                Object {
                  "$$computedPropsFromProxy": Object {
                    "buttonType": "default",
                    "style": Object {
                      "display": "block",
                    },
                  },
                  "$$parentTemplate": undefined,
                  "$$refForProxy": Object {},
                  "brick": "basic-bricks.general-button",
                  "events": Object {
                    "general.button.click": Object {
                      "action": "console.log",
                      "args": Array [
                        "source",
                        "\${EVENT}",
                      ],
                    },
                  },
                  "properties": Object {
                    "buttonType": "dashed",
                  },
                  "slots": Object {},
                },
                Object {
                  "$$computedPropsFromProxy": Object {},
                  "$$parentTemplate": Object {
                    "proxy": Object {
                      "events": Object {
                        "button.click": Object {
                          "ref": "button",
                          "refEvent": "general.button.click",
                        },
                      },
                      "methods": Object {
                        "tell": Object {
                          "ref": "button",
                          "refMethod": "tellStory",
                        },
                      },
                      "properties": Object {
                        "button": Object {
                          "ref": "button",
                          "refProperty": "buttonName",
                        },
                        "isDanger": Object {
                          "ref": "button",
                          "refTransform": Object {
                            "buttonType": "<% DATA.isDanger ? 'danger' : 'default' %>",
                            "style": Object {
                              "display": "<% DATA.isDanger ? 'inline' : 'block' %>",
                            },
                          },
                        },
                        "noGap": Object {
                          "ref": "micro-view",
                          "refProperty": "noGap",
                        },
                      },
                      "slots": Object {
                        "extraContent": Object {
                          "ref": "micro-view",
                          "refSlot": "content",
                        },
                        "tools": Object {
                          "ref": "micro-view",
                          "refPosition": 0,
                          "refSlot": "toolbar",
                        },
                      },
                    },
                    "proxyRefs": Map {
                      "button" => Object {},
                      "micro-view" => Object {},
                    },
                  },
                  "$$refForProxy": undefined,
                  "brick": "basic-bricks.brick-in-portal",
                  "portal": true,
                  "slots": Object {},
                },
                Object {
                  "brick": "div",
                  "properties": Object {
                    "textContent": "slotted extra content!",
                  },
                },
              ],
              "type": "bricks",
            },
            "toolbar": Object {
              "bricks": Array [
                Object {
                  "brick": "div",
                  "properties": Object {
                    "textContent": "slotted tools!",
                  },
                },
              ],
              "type": "bricks",
            },
          },
        },
      ],
      "type": "bricks",
    },
  },
}
`;

exports[`expandCustomTemplate should expandCustomTemplate 2`] = `
Object {
  "proxy": Object {
    "events": Object {
      "button.click": Object {
        "ref": "button",
        "refEvent": "general.button.click",
      },
    },
    "methods": Object {
      "tell": Object {
        "ref": "button",
        "refMethod": "tellStory",
      },
    },
    "properties": Object {
      "button": Object {
        "ref": "button",
        "refProperty": "buttonName",
      },
      "isDanger": Object {
        "ref": "button",
        "refTransform": Object {
          "buttonType": "<% DATA.isDanger ? 'danger' : 'default' %>",
          "style": Object {
            "display": "<% DATA.isDanger ? 'inline' : 'block' %>",
          },
        },
      },
      "noGap": Object {
        "ref": "micro-view",
        "refProperty": "noGap",
      },
    },
    "slots": Object {
      "extraContent": Object {
        "ref": "micro-view",
        "refSlot": "content",
      },
      "tools": Object {
        "ref": "micro-view",
        "refPosition": 0,
        "refSlot": "toolbar",
      },
    },
  },
  "proxyRefs": Map {
    "button" => Object {},
    "micro-view" => Object {},
  },
}
`;
